algorithms/src/std_algorithms/Kokkos_PartitioningOperations.hpp

```diff
@@ -0,0 +1,495 @@
+  PredType m_pred;
+
+  KOKKOS_FUNCTION
+  StdPartitionCopyFunctor(FirstFrom first_from, FirstDestTrue first_dest_true,
+                          FirstDestFalse first_dest_false, PredType pred)
+      : m_first_from(std::move(first_from)),
+        m_first_dest_true(std::move(first_dest_true)),
+        m_first_dest_false(std::move(first_dest_false)),
+        m_pred(std::move(pred)) {}
+
+  KOKKOS_FUNCTION
+  void operator()(const IndexType i, value_type& update,
+                  const bool final_pass) const {
+    const auto& myval = m_first_from[i];
+    if (final_pass) {
+      if (m_pred(myval)) {
+        m_first_dest_true[update.true_count_] = myval;
+      } else {
+        m_first_dest_false[update.false_count_] = myval;
+      }
+    }
+
+    if (m_pred(myval)) {
+      update.true_count_ += 1;
+    } else {
+      update.false_count_ += 1;
+    }
+  }
+
+  KOKKOS_FUNCTION
+  void init(value_type& update) const {
+    update.true_count_  = 0;
+    update.false_count_ = 0;
+  }
+
+  KOKKOS_FUNCTION
+  void join(volatile value_type& update,
+            volatile const value_type& input) const {
+    update.true_count_ += input.true_count_;
+    update.false_count_ += input.false_count_;
+  }
+};
+
+// ------------------------------------------
+// is_partitioned_impl
+// ------------------------------------------
+
+template <class ExecutionSpace, class IteratorType, class PredicateType>
+bool is_partitioned_impl(const std::string& label, const ExecutionSpace& ex,
+                         IteratorType first, IteratorType last,
+                         PredicateType pred) {
+  // true if all elements in the range [first, last) that satisfy
+  // the predicate "pred" appear before all elements that don't.
+  // Also returns true if [first, last) is empty.
+  // also true if all elements satisfy the predicate.
+
+  // we implement it by finding:
+  // - the max location where predicate is true  (max_loc_true)
+  // - the min location where predicate is false (min_loc_false)
+  // so the range is partitioned if max_loc_true < (min_loc_false)
+
+  // checks
+  Impl::static_assert_random_access_and_accessible(ex, first);
+  Impl::expect_valid_range(first, last);
+
+  // trivial case
+  if (first == last) {
+    return true;
+  }
+
+  // aliases
+  using index_type       = typename IteratorType::difference_type;
+  using reducer_type     = StdIsPartitioned<index_type, ExecutionSpace>;
+  using result_view_type = typename reducer_type::result_view_type;
+  using func_t =
+      StdIsPartitionedFunctor<IteratorType, reducer_type, PredicateType>;
+
+  // run
+  result_view_type result("is_partitioned_impl_result_view");
+  reducer_type reducer(result);
+  const auto num_elements = last - first;
+  ::Kokkos::parallel_reduce(label,
+                            RangePolicy<ExecutionSpace>(ex, 0, num_elements),
+                            func_t(first, reducer, pred), reducer);
+  ex.fence("Kokkos::is_partitioned: fence after operation");
```

Comment:
`create_mirror_view_and_copy` call below will synchronize.  You don't need this fence.

Reply:
thanks, yes good point, i have changed all similar instances